Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -1133,6 +1181,288 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+        if (deniedEntry.key->matchesAllHosts())
+            continue;
+
+        // Only revoked patterns that match the granted pattern need to be included. This limits
+        // the size of the exclude match patterns list to speed up processing.
+        if (!pattern.matchesPattern(deniedEntry.key, { WebExtensionMatchPattern::Options::IgnorePaths, WebExtensionMatchPattern::Options::MatchBidirectionally }))
+            continue;
+
+        [baseExcludeMatchPatternsSet addObjectsFromArray:deniedEntry.key->expandedStrings()];
+    }
+
+    auto allUserContentControllers = extensionController()->allUserContentControllers();
+
+    for (auto& injectedContentData : injectedContents) {
+        NSMutableSet<NSString *> *includeMatchPatternsSet = [NSMutableSet set];
+
+        for (auto& includeMatchPattern : injectedContentData.includeMatchPatterns) {
+            // Paths are not matched here since all we need to match at this point is scheme and host.
+            // The path matching will happen in WebKit when deciding to inject content into a frame.
+
+            // When the include pattern matches all hosts, we can generate a restricted patten here and skip
+            // the more expensive calls to matchesPattern() below since we know they will match.
+            if (includeMatchPattern->matchesAllHosts()) {
+                auto restrictedPattern = WebExtensionMatchPattern::getOrCreate(includeMatchPattern->scheme() ?: @"*", pattern.host() ?: @"*", includeMatchPattern->path() ?: @"/*");
+                if (!restrictedPattern)
+                    continue;
+
+                [includeMatchPatternsSet addObjectsFromArray:restrictedPattern->expandedStrings()];
+                continue;
+            }
+
+            // When deciding if injected content patterns match, we need to check bidirectionally.
+            // This allows an extension that requests *.wikipedia.org, to still inject content when
+            // it is granted more specific access to *.en.wikipedia.org.
+            if (!includeMatchPattern->matchesPattern(pattern, { WebExtensionMatchPattern::Options::IgnorePaths, WebExtensionMatchPattern::Options::MatchBidirectionally }))
+                continue;
+
+            // Pick the most restrictive match pattern by comparing unidirectionally to the granted origin pattern.
+            // If the include pattern still matches the granted origin pattern, it is not restrictive enough.
+            // In that case we need to use the include pattern scheme and path, but with the granted pattern host.
+            RefPtr restrictedPattern = includeMatchPattern.ptr();
+            if (includeMatchPattern->matchesPattern(pattern, { WebExtensionMatchPattern::Options::IgnoreSchemes, WebExtensionMatchPattern::Options::IgnorePaths }))
+                restrictedPattern = WebExtensionMatchPattern::getOrCreate(includeMatchPattern->scheme() ?: @"*", pattern.host() ?: @"*", includeMatchPattern->path() ?: @"/*");
```

Comment:
It looks like we use this structure a couple places (with defaults for missing scheme, host, and path). Could this be put into a helper method?

Reply:
I moved it inside `WebExtensionMatchPattern::getOrCreate()`